dag.go

```diff
@@ -102,6 +109,152 @@ func (m *Manifest) RootCID() cid.Cid {
+		PrevLabels[index] = path
+	}
+
+	lastLink := m.Links[len(m.Links)-1]
+	return &subDAGConverter{
+		Conversion:         map[int]int{},
+		PrevManifest:       m,
+		Manifest:           &Manifest{Nodes: []string{}, Links: [][2]int{}},
+		currentParentIndex: 0,
+		Parents:            []int{root},
+		LastBranchIndex:    lastLink[0],
+		PrevSizes:          prevInfo.Sizes,
+		Sizes:              []uint64{},
+		PrevLabels:         PrevLabels,
+		Labels:             map[string]int{},
+	}, nil
+}
+
+type subDAGConverter struct {
+	Conversion         map[int]int
+	PrevManifest       *Manifest
+	currentParentIndex int
+	Parents            []int
+	LastBranchIndex    int
+	Manifest           *Manifest
+	PrevSizes          []uint64
+	Sizes              []uint64
+	PrevLabels         map[int]string
+	Labels             map[string]int
+}
+
+func (s *subDAGConverter) Convert() {
+	for _, link := range s.PrevManifest.Links {
+		fromNode := link[0]
+		toNode := link[1]
+		currentParent := s.CurrentParent()
+		if fromNode < currentParent {
+			continue
+		}
+		if fromNode > currentParent {
+			for {
+				currentParent := s.NextParent()
+				if currentParent == -1 {
+					// have iterated through all the parents!
+					// this means our manifest is complete!
+					return
+				}
+				if currentParent == fromNode {
+					// break out of the forloop and continue
+					// with the Convert function
+					break
+				}
+				// keep iterating through the Parents array
+				continue
+			}
+		}
+		_, fromOk := s.Conversion[fromNode]
+		if !fromOk {
+			convertIndex := len(s.Manifest.Nodes)
+			id := s.PrevManifest.Nodes[fromNode]
+			s.Manifest.Nodes = append(s.Manifest.Nodes, id)
+			if s.PrevSizes != nil {
+				s.Sizes = append(s.Sizes, s.PrevSizes[fromNode])
+			}
+			if s.PrevLabels != nil {
+				path, ok := s.PrevLabels[fromNode]
+				if ok {
+					s.Labels[path] = convertIndex
+				}
+			}
+			s.Conversion[fromNode] = convertIndex
+		}
+		_, toOk := s.Conversion[toNode]
+		if !toOk {
+			convertIndex := len(s.Manifest.Nodes)
+			id := s.PrevManifest.Nodes[toNode]
+			s.Manifest.Nodes = append(s.Manifest.Nodes, id)
+			if s.PrevSizes != nil {
+				s.Sizes = append(s.Sizes, s.PrevSizes[toNode])
+			}
+			if s.PrevLabels != nil {
+				path, ok := s.PrevLabels[toNode]
+				if ok {
+					s.Labels[path] = convertIndex
+				}
+			}
+			s.Conversion[toNode] = convertIndex
+		}
+
+		newLink := [2]int{s.Conversion[fromNode], s.Conversion[toNode]}
+		s.Manifest.Links = append(s.Manifest.Links, newLink)
+		if toNode <= s.LastBranchIndex {
+			s.Parents = append(s.Parents, toNode)
+		}
+	}
+}
+
+func (s *subDAGConverter) CurrentParent() int {
+	return s.Parents[s.currentParentIndex]
+}
+
+func (s *subDAGConverter) NextParent() int {
+	s.currentParentIndex++
+	if s.currentParentIndex >= len(s.Parents) {
+		return -1
+	}
+	return s.CurrentParent()
+}
+
+// IDIndex returns the node index of the id
+func (m *Manifest) IDIndex(id string) (int, error) {
```

Comment:
I'm sorry, I take back my earlier suggestion, seeing now how this is being used. Returning -1 and no error is totally fine, since there's no real info being gained when the index isn't found. That applies to any case where the method is just going the standard "find" algorithm, and the error contains nothing interesting. Apologies for going down the wrong path at first.